### PR TITLE
[UPDATED PROTOTYPE] Use dynamo fake tensor mode in aot_autograd, move aot_autograd compilation to lowering time

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89773
* __->__ #89672

After all of the preparatory commits, this is a subset of the
changes in https://github.com/pytorch/pytorch/pull/89392 that actually
change us to propagating fake tensors to backends.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_dynamo/optimizations/distributed.py [hunk 1]
```diff
@@ -236,42 +240,61 @@ def forward(self, *args):
...
                     args, kwargs = self.fetch_args_kwargs_from_env(n)
+                    new_args = []
+                    assert fake_mode
+                    for arg in args:
+                        if isinstance(arg, torch.Tensor) and not isinstance(
+                            arg, torch._subclasses.FakeTensor
+                        ):
+                            new_args.append(fake_mode.from_tensor(arg))
+                        else:
+                            new_args.append(arg)
```
Comment (voznesenskym) [inline torch/_dynamo/optimizations/distributed.py:267]: I kept meaning to rewrite this as
```suggestion
                    def convert(arg):
                        if isinstance(arg, torch.Tensor) and not isinstance(
                            arg, torch._subclasses.FakeTensor
                        ):
                            return arg
                        return arg
                     
                    new_args = [convert(arg) for arg in args]
```

We also dot this in enough places that a util my be warranted...

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -398,6 +398,10 @@ def clone_tensor(x):
 
 def clone_input(x):
     """copy while preserving strides"""
+    # TODO: this is questionable
+    if isinstance(x, torch._subclasses.FakeTensor):
+        # this func fails on fake tensors in __torch_dispatch__
+        return x
```
Comment (voznesenskym) [inline torch/_dynamo/utils.py:404]: We had umbrage with this in the source PR, did we ever get to talk about what a better course might be?

File: torch/_dynamo/utils.py [hunk 3]
```diff
@@ -1144,3 +1148,19 @@ def get_real_value(node, output_graph):
     except RuntimeError as e:
         raise TorchRuntimeError() from e
     return real_value
+
+def fake_mode_from_tensors(inputs: List[Any]):
```
Comment (voznesenskym) [inline torch/_dynamo/utils.py:1152]: ah and here it is

**Review by voznesenskym — APPROVED** [review]

**Review by Chillee — COMMENTED** [review]
Tentatively LGTM, might need to read more closely.

**Comment by voznesenskym** [comment]
Gonna look at tests on this one.

**Comment by voznesenskym** [comment]
Oh this is missing the C++ fix @ezyang

File: torch/_dynamo/optimizations/distributed.py [hunk 4]
```diff
@@ -236,42 +240,61 @@ def forward(self, *args):
...
                     if n.op == "call_module":
-                        submod = self.fetch_attr(n.target)
-                        log.debug(f"\n---{n.target} graph---\n" + str(submod.graph))
-                        compiled_submod = self.compile_submod(submod, args, kwargs)
+                        real_mod = self.fetch_attr(n.target)
+                        if fake_mode:
+                            fake_submod = deepcopy_to_fake_tensor(real_mod, fake_mode)
+                        else:
+                            fake_submod = real_mod
+                            pass
```
Comment (wconstab) [inline torch/_dynamo/optimizations/distributed.py:281]: when is `not fake_mode`?  and why this `pass`?

File: torch/_dynamo/output_graph.py [hunk 5]
```diff
@@ -540,15 +528,27 @@ def call_user_compiler(self, gm: fx.GraphModule) -> CompiledFn:
...
+    def fake_example_inputs(self) -> List[torch.Tensor]:
+        result = []
+        for arg in self.graphargs:
+            example = arg.get_fake_examples()
+            if example is not None:
+                result.extend(example)
+            else:
+                # Fallback, in case fake_tensor was not set
+                # Particularly for graph args that are not tensors
+                result.extend(arg.get_examples())
```
Comment (bdhirsh) [inline torch/_dynamo/output_graph.py:549]: Two questions about this comment:

(1) The return type of this func is `List[torch.Tensor]`, is that wrong? (comment implies we have to worry about non-tensor graph args)

(2) Should this function maintain the invariant that it never returns real tensors, and assert here that `get_examples()` in the fallback only returns non-tensors?

File: torch/_dynamo/variables/builder.py [hunk 6]
```diff
@@ -551,10 +565,7 @@ def wrap_tensor(self, value: torch.Tensor):
                 # guards=self.make_guards(GuardBuilder.TENSOR_MATCH),
             )
         else:
-            if not is_constant_source(self.get_source()):
-                self.tx.output.graphargs.append(
-                    GraphArg(self.get_source(), value, False)
-                )
+            graph_arg = None
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:568]: There's no need to define graph_arg here

**Comment by ezyang** [comment]
obsolete

**Pull request closed without merging**